File: java/parser/src/main/java/gov/lbl/parser/parser/StatementVisitor.java
package gov.lbl.parser.parser;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Assignment_statement;
import gov.lbl.parser.domain.Assignment_with_function_call_statement;
import gov.lbl.parser.domain.Comment;
import gov.lbl.parser.domain.Component_reference;
import gov.lbl.parser.domain.Expression;
import gov.lbl.parser.domain.For_statement;
import gov.lbl.parser.domain.Function_call_args;
import gov.lbl.parser.domain.Function_call_statement;
import gov.lbl.parser.domain.If_statement;
import gov.lbl.parser.domain.Output_expression_list;
import gov.lbl.parser.domain.Statement;
import gov.lbl.parser.domain.When_statement;
import gov.lbl.parser.domain.While_statement;

public class StatementVisitor extends modelicaBaseVisitor<Statement> {
    @Override
    public Statement visitStatement(modelicaParser.StatementContext ctx) {
        Component_referenceVisitor component_referenceVisitor = new Component_referenceVisitor();
        Component_reference component_reference = ctx.component_reference() == null ? null : ctx.component_reference().accept(component_referenceVisitor);

        ExpressionVisitor expressionVisitor = new ExpressionVisitor();
        Expression value = ctx.expression() == null ? null : ctx.expression().accept(expressionVisitor);

        Function_call_argsVisitor function_call_argsVisitor = new Function_call_argsVisitor();
        Function_call_args function_call_args = ctx.function_call_args() == null ? null : ctx.function_call_args().accept(function_call_argsVisitor);
        
        Output_expression_listVisitor output_expression_listVisitor = new Output_expression_listVisitor();
        Output_expression_list output_expression_list = ctx.output_expression_list() == null ? null : ctx.output_expression_list().accept(output_expression_listVisitor);

        Assignment_statement assignment_statement = null;
        Function_call_statement function_call_statement = null;
        Assignment_with_function_call_statement assignment_with_function_call_statement = null;
        if (component_reference != null && value != null) {
            assignment_statement = new Assignment_statement(component_reference, value);
        }
        else if (component_reference != null && function_call_args != null && output_expression_list != null) {
            assignment_with_function_call_statement = new Assignment_with_function_call_statement(output_expression_list, component_reference, function_call_args);
        }
        else if (component_reference != null && function_call_args != null) {
            function_call_statement = new Function_call_statement(component_reference, function_call_args);
        }

        Boolean is_break = ctx.BREAK() == null ? false : true;
        Boolean is_return = ctx.RETURN() == null ? false : true;
    
        If_statementVisitor if_statementVisitor = new If_statementVisitor();
        If_statement if_statement = ctx.if_statement() == null ? null : ctx.if_statement().accept(if_statementVisitor);

        For_statementVisitor for_statementVisitor = new For_statementVisitor();
        For_statement for_statement = ctx.for_statement() == null ? null : ctx.for_statement().accept(for_statementVisitor);
        
        While_statementVisitor while_statementVisitor = new While_statementVisitor();
        While_statement while_statement = ctx.while_statement() == null ? null : ctx.while_statement().accept(while_statementVisitor);
        
        When_statementVisitor when_statementVisitor = new When_statementVisitor();
        When_statement when_statement = ctx.when_statement() == null ? null : ctx.when_statement().accept(when_statementVisitor);
        
        CommentVisitor commentVisitor = new CommentVisitor();
        Comment comment = ctx.comment() == null ? null : ctx.comment().accept(commentVisitor);
        
        return new Statement(assignment_statement, function_call_statement, assignment_with_function_call_statement, is_break, is_return, if_statement, for_statement, while_statement, when_statement, comment);
    }
}

File: jsParser/parser/StatementVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Statement = require('../domain/Statement')
const Assignment_statement = require('../domain/Assignment_statement')
const Assignment_with_function_call_statement = require('../domain/Assignment_with_function_call_statement')
const Function_call_statement = require('../domain/Function_call_statement')

const ExpressionVisitor = require('./ExpressionVisitor')
const Component_referenceVisitor = require('./Component_referenceVisitor')
const Function_call_argsVisitor = require('./Function_call_argsVisitor')
const Output_expression_listVisitor = require('./Output_expression_listVisitor')
const If_statementVisitor = require('./If_statementVisitor')
const For_statementVisitor = require('./For_statementVisitor')
const While_statementVisitor = require('./While_statementVisitor')
const When_statementVisitor = require('./When_statementVisitor')
const CommentVisitor = require('./CommentVisitor')

class StatementVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitStatement (ctx) {
    let assignment_statement = null
    let assignment_with_function_call_statement = null
    let component_reference = null
    let value = null
    let function_call_statement = null
    let function_call_args = null
    const is_break = !!ctx.BREAK()
    const is_return = !!ctx.RETURN()
    let if_statement = null
    let for_statement = null
    let while_statement = null
    let when_statement = null
    let comment = null
    let output_expression_list = null

    if (ctx.component_reference()) {
      const component_referenceVisitor = new Component_referenceVisitor.Component_referenceVisitor()
      component_reference = component_referenceVisitor.visitComponent_reference(ctx.component_reference())
    }
    if (ctx.expression()) {
      const expressionVisitor = new ExpressionVisitor.ExpressionVisitor()
      value = expressionVisitor.visitExpression(ctx.expression())
    }
    if (ctx.function_call_args()) {
      const function_call_argsVisitor = new Function_call_argsVisitor.Function_call_argsVisitor()
      function_call_args = function_call_argsVisitor.visitFunction_call_args(ctx.function_call_args())
    }
    if (ctx.output_expression_list()) {
      const output_expression_listVisitor = new Output_expression_listVisitor.Output_expression_listVisitor()
      output_expression_list = output_expression_listVisitor.visitOutput_expression_list(ctx.output_expression_list())
    }

    if (component_reference != null && value != null) {
      assignment_statement = new Assignment_statement.Assignment_statement(component_reference, value)
    } else if (component_reference != null && function_call_args != null && output_expression_list != null) {
      assignment_with_function_call_statement = new Assignment_with_function_call_statement.Assignment_with_function_call_statement(output_expression_list, component_reference, function_call_args)
    } else if (component_reference != null && function_call_args != null) {
      function_call_statement = new Function_call_statement.Function_call_statement(component_reference, function_call_args)
    }
    if (ctx.if_statement()) {
      const if_statementVisitor = new If_statementVisitor.If_statementVisitor()
      if_statement = if_statementVisitor.visitIf_statement(ctx.if_statement())
    }
    if (ctx.for_statement()) {
      const for_statementVisitor = new For_statementVisitor.For_statementVisitor()
      for_statement = for_statementVisitor.visitFor_statement(ctx.for_statement())
    }
    if (ctx.while_statement()) {
      const while_statementVisitor = new While_statementVisitor.While_statementVisitor()
      while_statement = while_statementVisitor.visitWhile_statement(ctx.while_statement())
    }
    if (ctx.when_statement()) {
      const when_statementVisitor = new When_statementVisitor.When_statementVisitor()
      when_statement = when_statementVisitor.visitWhen_statement(ctx.when_statement())
    }
    if (ctx.comment()) {
      const commentVisitor = new CommentVisitor.CommentVisitor()
      comment = commentVisitor.visitComment(ctx.comment())
    }

    return new Statement.Statement(assignment_statement, function_call_statement, assignment_with_function_call_statement, is_break,
      is_return, if_statement, for_statement, while_statement, when_statement, comment)
  }
};

StatementVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitStatement = this.visitStatement
exports.StatementVisitor = StatementVisitor
